File: neml2/csrc/aoti/log.cpp
// Copyright 2024, UChicago Argonne, LLC
// All Rights Reserved
// Software Name: NEML2 -- the New Engineering material Model Library, version 2
// By: Argonne National Laboratory
// OPEN SOURCE LICENSE (MIT)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include "neml2/csrc/aoti/log.h"
#include "neml2/csrc/aoti/Exception.h"

#include <cctype>
#include <cstdlib>
#include <iostream>
#include <mutex>
#include <optional>
#include <string>

namespace neml2::aoti::log
{
namespace
{
constexpr int NCHAN = 6;

int
idx(Channel c)
{
  return static_cast<int>(c);
}

// The default sink: split by severity so normal output is pipeable on stdout and
// diagnostics/warnings stay out of the way on stderr. One flushed line each so a
// nested (linear) residual never lags the Newton iteration it belongs to.
void
default_sink(Level lvl, const std::string & line)
{
  if (lvl == Level::Info)
    std::cout << line << std::endl;
  else
    std::cerr << line << std::endl;
}

// Process-global config. The two layers are `std::optional` per channel plus an
// `all` baseline; effective resolution walks env[c] -> env_all -> prog[c] ->
// prog_all -> built-in default.
struct State
{
  std::optional<Level> env[NCHAN];
  std::optional<Level> env_all;
  std::optional<Level> prog[NCHAN];
  std::optional<Level> prog_all;
  Sink sink = &default_sink;
  bool env_applied = false;
  std::mutex mtx;
};

State &
state()
{
  static State s;
  return s;
}

std::string
trim_lower(const std::string & s)
{
  std::size_t a = 0, b = s.size();
  while (a < b && std::isspace(static_cast<unsigned char>(s[a])))
    ++a;
  while (b > a && std::isspace(static_cast<unsigned char>(s[b - 1])))
    --b;
  std::string out = s.substr(a, b - a);
  for (auto & ch : out)
    ch = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
  return out;
}

// Clear + re-read the env layer from NEML2_LOGS. Caller holds the lock.
void
apply_env_locked(State & s)
{
  for (auto & e : s.env)
    e.reset();
  s.env_all.reset();
  s.env_applied = true;

  const char * raw = std::getenv("NEML2_LOGS");
  if (!raw)
    return;

  const std::string spec(raw);
  std::size_t pos = 0;
  while (pos <= spec.size())
  {
    const std::size_t comma = spec.find(',', pos);
    const std::string entry =
        spec.substr(pos, comma == std::string::npos ? std::string::npos : comma - pos);
    pos = (comma == std::string::npos) ? spec.size() + 1 : comma + 1;

    const std::size_t eq = entry.find('=');
    if (eq == std::string::npos)
      continue;
    const std::string chan = trim_lower(entry.substr(0, eq));
    const std::string lvl = trim_lower(entry.substr(eq + 1));
    Level L;
    if (!parse_level(lvl, L))
      throw FatalError("NEML2_LOGS: unknown level '" + lvl + "' in entry '" + trim_lower(entry) +
                       "' (expected silent|warning|info|debug or 0..3)");
    if (chan == "all")
      s.env_all = L;
    else
    {
      Channel C;
      if (!parse_channel(chan, C))
        throw FatalError("NEML2_LOGS: unknown channel '" + chan +
                         "' (expected all|newton|linear|substep|model|tensor|driver)");
      s.env[idx(C)] = L;
    }
  }
}

Level
effective_locked(State & s, Channel c)
{
  if (!s.env_applied)
    apply_env_locked(s);
  const int i = idx(c);
  if (s.env[i])
    return *s.env[i];
  if (s.env_all)
    return *s.env_all;
  if (s.prog[i])
    return *s.prog[i];
  if (s.prog_all)
    return *s.prog_all;
  return Level::Warning;
}
} // namespace

bool
enabled(Channel channel, Level level)
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  return static_cast<int>(effective_locked(s, channel)) >= static_cast<int>(level);
}

Level
effective_level(Channel channel)
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  return effective_locked(s, channel);
}

std::string
format(Channel channel, const std::string & message)
{
  // Pad the channel to the width of the longest name ("substep") so the prefix
  // is column-aligned across channels: "[neml2:newton ] ..." / "[neml2:substep] ...".
  constexpr std::size_t field = 7;
  std::string name = channel_name(channel);
  if (name.size() < field)
    name.append(field - name.size(), ' ');
  return "[neml2:" + name + "] " + message;
}

void
emit(Channel channel, Level level, const std::string & message)
{
  // Copy the sink out under the lock, then call it unlocked: a custom sink may
  // take the GIL (the Python-forwarding sink) or do I/O, and holding our mutex
  // across that would risk a lock-ordering deadlock with the GIL.
  Sink sink_copy;
  bool on = false;
  {
    auto & s = state();
    std::lock_guard<std::mutex> lock(s.mtx);
    on = static_cast<int>(effective_locked(s, channel)) >= static_cast<int>(level);
    if (on)
      sink_copy = s.sink;
  }
  if (on && sink_copy)
    sink_copy(level, format(channel, message));
}

void
begin_solve(Channel channel, const std::string & label)
{
  emit(channel, Level::Info, "---- begin " + label + " ----");
}

void
end_solve(Channel channel, const std::string & label)
{
  emit(channel, Level::Info, "---- end " + label + " ----");
}

void
set_default_level(Channel channel, Level level)
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  s.prog[idx(channel)] = level;
}

void
set_default_level(Level level)
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  s.prog_all = level;
}

void
reset_defaults()
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  for (auto & p : s.prog)
    p.reset();
  s.prog_all.reset();
}

void
set_sink(Sink sink)
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  s.sink = sink ? std::move(sink) : Sink(&default_sink);
}

void
reset_sink()
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  s.sink = &default_sink;
}

void
apply_env()
{
  auto & s = state();
  std::lock_guard<std::mutex> lock(s.mtx);
  apply_env_locked(s);
}

const char *
channel_name(Channel channel)
{
  switch (channel)
  {
    case Channel::Newton:
      return "newton";
    case Channel::Linear:
      return "linear";
    case Channel::Substep:
      return "substep";
    case Channel::Model:
      return "model";
    case Channel::Tensor:
      return "tensor";
    case Channel::Driver:
      return "driver";
  }
  throw FatalError("neml2::aoti::log::channel_name: unrecognized Channel enum value");
}

const char *
level_name(Level level)
{
  switch (level)
  {
    case Level::Silent:
      return "silent";
    case Level::Warning:
      return "warning";
    case Level::Info:
      return "info";
    case Level::Debug:
      return "debug";
  }
  throw FatalError("neml2::aoti::log::level_name: unrecognized Level enum value");
}

bool
parse_channel(const std::string & s, Channel & out)
{
  if (s == "newton")
    out = Channel::Newton;
  else if (s == "linear")
    out = Channel::Linear;
  else if (s == "substep")
    out = Channel::Substep;
  else if (s == "model")
    out = Channel::Model;
  else if (s == "tensor")
    out = Channel::Tensor;
  else if (s == "driver")
    out = Channel::Driver;
  else
    return false;
  return true;
}

bool
parse_level(const std::string & s, Level & out)
{
  if (s == "silent" || s == "off" || s == "none" || s == "0")
    out = Level::Silent;
  else if (s == "warning" || s == "warn" || s == "1")
    out = Level::Warning;
  else if (s == "info" || s == "2")
    out = Level::Info;
  else if (s == "debug" || s == "3")
    out = Level::Debug;
  else
    return false;
  return true;
}
} // namespace neml2::aoti::log

File: neml2/log.py
# Copyright 2024, UChicago Argonne, LLC
# All Rights Reserved
# Software Name: NEML2 -- the New Engineering material Model Library, version 2
# By: Argonne National Laboratory
# OPEN SOURCE LICENSE (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

"""Central verbosity / logging control for NEML2 -- shared by all six routes.

Every neml2-emitted *library diagnostic* (solver traces, substepping, model /
equation-system debug, driver notices) is gated through one config store. The
same store lives in ``libneml2.so`` (so the C++ solvers obey it directly, whether
reached from C++ or from Python) and is surfaced here; this module is a thin
client plus the output sink.

Control it two ways, resolved highest-first:

1. the ``NEML2_LOGS`` environment variable (the end-user override) --
   ``TORCH_LOGS``-style, e.g. ``NEML2_LOGS="newton=info,substep=debug"`` or
   ``NEML2_LOGS="all=warning"``;
2. programmatic defaults set by a downstream app via :func:`set_default_level`.

Then the built-in default (``warning``) fills the rest. Levels, least to most
verbose: ``silent < warning < info < debug``. Channels: ``newton``, ``linear``,
``substep``, ``model``, ``tensor``, ``driver`` (plus ``all`` as a baseline).

Output goes to ``sys.stdout`` for ``info`` and ``sys.stderr`` for
``warning`` / ``debug`` by default; redirect with :func:`set_stream`,
:func:`set_streams`, or :func:`set_sink`.
"""

from __future__ import annotations

import sys
from collections.abc import Callable
from typing import Protocol, cast

#: Level names, ordered least-to-most verbose (mirrors ``neml2/csrc/aoti/log.h``).
LEVELS = ("silent", "warning", "info", "debug")
#: Diagnostic channels (mirrors ``neml2/csrc/aoti/log.h``).
CHANNELS = ("newton", "linear", "substep", "model", "tensor", "driver")

# Canonical level constants for callers who prefer symbols over string literals.
SILENT = "silent"
WARNING = "warning"
INFO = "info"
DEBUG = "debug"

_LEVEL_ORD = {name: i for i, name in enumerate(LEVELS)}
# Accepted aliases -> canonical level (mirrors C++ parse_level).
_LEVEL_ALIASES = {
    "off": "silent",
    "none": "silent",
    "warn": "warning",
    "0": "silent",
    "1": "warning",
    "2": "info",
    "3": "debug",
}


def _norm_level(level: str) -> str:
    """Canonical level name, or raise ``ValueError`` on an unrecognized token."""
    key = str(level).strip().lower()
    key = _LEVEL_ALIASES.get(key, key)
    if key not in _LEVEL_ORD:
        raise ValueError(
            f"unknown log level {level!r} (expected one of {', '.join(LEVELS)} or 0..3)"
        )
    return key


def _norm_channel(channel: str, *, allow_all: bool = False) -> str:
    """Canonical channel name, or raise ``ValueError`` on an unrecognized token."""
    key = str(channel).strip().lower()
    if key == "all" and allow_all:
        return "all"
    if key not in CHANNELS:
        extra = "all, " if allow_all else ""
        raise ValueError(
            f"unknown log channel {channel!r} (expected one of {extra}{', '.join(CHANNELS)})"
        )
    return key


# ---------------------------------------------------------------------------
# Backend: the one store in libneml2.so, reached through the _aoti binding. When
# the compiled extension is unavailable (a partial build), fall back to a pure-
# Python store so Python-only diagnostics (and the AOTI-import-failure advisory)
# still function.
# ---------------------------------------------------------------------------
try:
    from neml2.aoti._aoti import log as _backend  # type: ignore[attr-defined]
except Exception:  # noqa: BLE001 (any import failure -> degrade to the Python store)
    _backend = None


class _PyStore:
    """Pure-Python mirror of the C++ store for partial builds (no ``_aoti``)."""

    def __init__(self) -> None:
        self._prog: dict[str, str] = {}
        self._prog_all: str | None = None
        self._env: dict[str, str] = {}
        self._env_all: str | None = None
        self._applied = False

    def apply_env(self) -> None:
        import os  # noqa: PLC0415

        self._env = {}
        self._env_all = None
        self._applied = True
        spec = os.environ.get("NEML2_LOGS")
        if not spec:
            return
        for entry in spec.split(","):
            if "=" not in entry:
                continue
            chan, _, lvl = entry.partition("=")
            level = _norm_level(lvl)
            chan = _norm_channel(chan, allow_all=True)
            if chan == "all":
                self._env_all = level
            else:
                self._env[chan] = level

    def effective_level(self, channel: str) -> str:
        if not self._applied:
            self.apply_env()
        if channel in self._env:
            return self._env[channel]
        if self._env_all is not None:
            return self._env_all
        if channel in self._prog:
            return self._prog[channel]
        if self._prog_all is not None:
            return self._prog_all
        return "warning"

    def enabled(self, channel: str, level: str) -> bool:
        return _LEVEL_ORD[self.effective_level(channel)] >= _LEVEL_ORD[level]

    def emit(self, channel: str, level: str, message: str) -> None:
        # Pad the channel to the longest name ("substep") for column alignment,
        # matching the C++ store's format().
        if self.enabled(channel, level):
            _forward(level, f"[neml2:{channel:<7}] {message}")

    def set_default_level(self, channel: str, level: str) -> None:
        if channel == "all":
            self._prog_all = level
        else:
            self._prog[channel] = level

    def reset_defaults(self) -> None:
        self._prog = {}
        self._prog_all = None


class _StoreProto(Protocol):
    def enabled(self, channel: str, level: str) -> bool: ...
    def effective_level(self, channel: str) -> str: ...
    def emit(self, channel: str, level: str, message: str) -> None: ...
    def set_default_level(self, channel: str, level: str) -> None: ...
    def reset_defaults(self) -> None: ...
    def apply_env(self) -> None: ...


#: The active store: the C++ one (via _aoti) when available, else the Python
#: fallback. Both expose the same method surface.
_store: _StoreProto = cast("_StoreProto", _backend if _backend is not None else _PyStore())


# ---------------------------------------------------------------------------
# Output sink. A single forwarding function is installed into the C++ store at
# import; it reads the module-level stream config at call time so redirection
# (notebooks, pytest capsys, set_stream) takes effect immediately.
# ---------------------------------------------------------------------------
_user_sink: Callable[[str, str], None] | None = None
_out = None  # None -> sys.stdout looked up dynamically
_err = None  # None -> sys.stderr looked up dynamically


def _forward(level: str, line: str) -> None:
    """The installed sink: route one already-formatted line by level."""
    if _user_sink is not None:
        _user_sink(level, line)
        return
    if level == "info":
        stream = _out if _out is not None else sys.stdout
    else:
        stream = _err if _err is not None else sys.stderr
    print(line, file=stream, flush=True)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------
def enabled(channel: str, level: str) -> bool:
    """Whether ``channel`` would emit at ``level`` (guard expensive log lines)."""
    channel = _norm_channel(channel)
    level = _norm_level(level)
    return bool(_store.enabled(channel, level))


def effective_level(channel: str) -> str:
    """The resolved level of ``channel`` (env > programmatic > built-in default)."""
    channel = _norm_channel(channel)
    return str(_store.effective_level(channel))


def emit(channel: str, level: str, message: str) -> None:
    """Emit ``message`` on ``channel`` at ``level`` iff enabled.

    The line is formatted as ``[neml2:<channel>] <message>`` and routed through
    the configured sink -- identical to a line the C++ solvers emit.
    """
    channel = _norm_channel(channel)
    level = _norm_level(level)
    _store.emit(channel, level, message)


def set_default_level(channel: str, level: str) -> None:
    """Set the programmatic default level for one channel (or ``"all"``).

    A downstream app calls this to pick its baseline verbosity; an ``NEML2_LOGS``
    entry for the same channel still overrides it.
    """
    channel = _norm_channel(channel, allow_all=True)
    level = _norm_level(level)
    _store.set_default_level(channel, level)


def reset_defaults() -> None:
    """Clear all programmatic defaults (revert to env > built-in default)."""
    _store.reset_defaults()


def reload() -> None:
    """Re-read ``NEML2_LOGS`` into the env layer (e.g. after changing it in a test)."""
    # Validate in Python first so a malformed spec raises a consistent ValueError
    # regardless of backend.
    import os  # noqa: PLC0415

    spec = os.environ.get("NEML2_LOGS")
    if spec:
        for entry in spec.split(","):
            if "=" not in entry:
                continue
            chan, _, lvl = entry.partition("=")
            _norm_level(lvl)
            _norm_channel(chan, allow_all=True)
    _store.apply_env()


def set_stream(stream) -> None:
    """Route *all* neml2 log output to a single stream (both info and warn/debug)."""
    global _out, _err, _user_sink
    _out = stream
    _err = stream
    _user_sink = None


def set_streams(out, err) -> None:
    """Route ``info`` to ``out`` and ``warning`` / ``debug`` to ``err``."""
    global _out, _err, _user_sink
    _out = out
    _err = err
    _user_sink = None


def set_sink(sink: Callable[[str, str], None] | None) -> None:
    """Install a custom sink ``sink(level_name, formatted_line)`` for all output.

    Pass ``None`` to restore the default stdout/stderr split.
    """
    global _user_sink
    if sink is None:
        reset_sink()
        return
    _user_sink = sink


def reset_sink() -> None:
    """Restore the default sink: ``info`` -> stdout, ``warning`` / ``debug`` -> stderr."""
    global _out, _err, _user_sink
    _out = None
    _err = None
    _user_sink = None


# Install the forwarding sink into the C++ store and prime the env layer so that
# all neml2 output -- including lines emitted from C++ -- flows through Python's
# streams (and is captured by notebooks / pytest).
if _backend is not None:
    _backend.set_sink(_forward)
reload()

__all__ = [
    "LEVELS",
    "CHANNELS",
    "SILENT",
    "WARNING",
    "INFO",
    "DEBUG",
    "enabled",
    "effective_level",
    "emit",
    "set_default_level",
    "reset_defaults",
    "reload",
    "set_stream",
    "set_streams",
    "set_sink",
    "reset_sink",
]
